pkg/policy/api/rule.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2016-2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"github.com/cilium/cilium/pkg/labels"
+)
+
+// Rule is a policy rule which must be applied to all endpoints which match the
+// labels contained in the endpointSelector
+//
+// Multiple rules can select the same endpoint in which case all rules are
+// accumulated additively
+//
+// The rule is split into an ingress section which contains all rules
+// applicable at ingress, and an egress section applicable at egress.
+//
+// Either ingres, egress, or both can be provided. If both ingress and egress
+// are omitted, the rule has no effect.
+type Rule struct {
+	// EndpointSelector selects all endpoints which should be subject to
+	// this rule. Cannot be empty.
+	EndpointSelector EndpointSelector `json:"selector"`
+
+	// Ingress contains all rule types which can be applied at ingress
+	//
+	// If omitted, no rules will be applied at ingress via this rule.
+	//
+	// +optional
+	Ingress *IngressRule `json:"ingress,omitempty"`
+
+	// Egress contains all rule types which can be applied at egress
+	//
+	// If omitted, no rules will be applied at egress via this rule
+	//
+	// +optional
+	Egress *EgressRule `json:"egress,omitempty"`
+
+	// Labels is a list of optional strings which can be used to
+	// re-identify the rule or to store metadata. It is possible to lookup
+	// or delete strings based on labels. Labels are not required to be
+	// unique, multiple rules can have overlapping or identical labels.
+	//
+	// +optional
+	Labels labels.LabelArray `json:"labels,omitempty"`
+
+	// Comment is a free form string, it can be used by the creator of the
+	// rule to store human readable explanation of the purpose of this
+	// rule. Rules cannot be identified by comment.
+	//
+	// +optional
+	Comment string `json:"comment,omitempty"`
+}
+
+// IngressRule contains all rule types which can be applied at ingress,
+// i.e. network traffic that originates outside of the endpoint and
+// is entering the endpoint selected by the endpointSelector.
+//
```

Comment:
What you say is currently not true. The `FromEndpoints` is in no way correlated to the `ToPorts`. We can possibly support that in the future but it is a substantial change. It also means that we can no longer map a single k8s `NetworkPolicy` to a single Cilium rule which would complicate usage quite a bit.